File: src/windows-native/LanMessenger.Tests/StressTests.cs
using LanMessenger.Core.Crypto;
using LanMessenger.Core.Persistence;
using LanMessenger.Core.Protocol;
using LanMessenger.Core.Services;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using NSec.Cryptography;
using System.Text;
using System.Text.Json;

namespace LanMessenger.Tests;

// Stress and regression tests for the LAN Messenger protocol stack.
//
// Goals
// -----
// • Verify that high-frequency concurrent operations (logging, crypto, frame
//   codec) produce correct results and never crash or dead-lock.
// • Verify that the rotation budget is maintained under continuous write load.
// • Ensure UI-thread safety is maintained (no marshalling exceptions from
//   background operations).
// • Catch common regressions: message-status downgrade, history truncation,
//   frame-codec round-trip correctness at boundary sizes.
//
// All tests run as part of the MSTest suite and in CI via pr-checks.yml and
// integration-test.yml.  No real sockets or persistent disk state is used
// outside the temp-directory overrides.
[TestClass]
public sealed class StressTests
{
    private string _tempDir = "";
    private long   _savedMaxBytes;
    private int    _savedMaxArchives;

    [TestInitialize]
    public void Setup()
    {
        _tempDir = Path.Combine(Path.GetTempPath(), $"StressTests-{Guid.NewGuid():N}");
        Directory.CreateDirectory(_tempDir);
        LanLogger.TestLogDirectoryOverride = _tempDir;
        LanLogger._TestResetHeaderFlag();
        _savedMaxBytes    = LanLogger.MaxBytes;
        _savedMaxArchives = LanLogger.MaxArchives;
    }

    [TestCleanup]
    public void Teardown()
    {
        LanLogger.MaxBytes    = _savedMaxBytes;
        LanLogger.MaxArchives = _savedMaxArchives;
        LanLogger.TestLogDirectoryOverride = null;
        LanLogger._TestResetHeaderFlag();
        try { Directory.Delete(_tempDir, recursive: true); } catch { }
    }

    // ── Logger stress ─────────────────────────────────────────────────────────────

    /// 16 threads × 200 writes across all logging methods = 3 200 total writes.
    /// Verifies thread-safety: no crashes, no corrupted lines.
    [TestMethod]
    [Timeout(30_000)]
    public void HighConcurrencyAcrossAllChannels()
    {
        LanLogger.MaxBytes    = 512 * 1024;   // 512 KiB — allow rotation mid-test
        LanLogger.MaxArchives = 3;

        const int threads = 16;
        const int writes  = 200;

        var tasks = Enumerable.Range(0, threads).Select(t => Task.Run(() =>
        {
            for (int i = 0; i < writes; i++)
            {
                // LanLogger has Info/Warn/Error/FileTransfer/Screenshot/Peer.
                // Structured-event helpers that don't exist (Discovery, Crypto,
                // UI, Retry) are expressed as plain Info calls.
                switch (t % 8)
                {
                    case 0: LanLogger.Info("Stress", $"t{t} i{i}");                                    break;
                    case 1: LanLogger.FileTransfer("chunk", transferId: $"t{t}", bytesSent: i * 1024L); break;
                    case 2: LanLogger.Screenshot("frame", widthPx: 1920, heightPx: 1080);               break;
                    case 3: LanLogger.Peer("ping", peer: $"10.0.{t}.1");                                break;
                    case 4: LanLogger.Info("Discovery", $"beacon_sent interfaces={t}");                  break;
                    case 5: LanLogger.Info("Crypto", "derive algorithm=X25519");                         break;
                    case 6: LanLogger.Info("UI", "frame_update screen=chat");                            break;
                    default: LanLogger.Info("Retry", $"retry subsystem=Transfer attempt={i}");           break;
                }
            }
        })).ToArray();

        Task.WaitAll(tasks);

        // The single active log must exist and be non-empty after the writes.
        if (File.Exists(LanLogger.LogPath))
            Assert.IsTrue(new FileInfo(LanLogger.LogPath).Length > 0,
                "active log should not be empty after concurrent writes");
    }

    /// Continuously rotates the log; verifies archive count stays within
    /// MaxArchives and the active log is bounded.
    [TestMethod]
    public void RotationBudgetMaintainedUnderLoad()
    {
        LanLogger.MaxBytes    = 256;
        LanLogger.MaxArchives = 2;

        for (int i = 0; i < 500; i++)
            LanLogger.FileTransfer("chunk", transferId: "rot",
                filename: $"file_{i}.bin", size: (long)i * 64);

        // LanLogger rotates the single client.log into client.N.log.gz archives.
        var archives = Directory.GetFiles(_tempDir, "client.*.log.gz");
        Assert.IsTrue(archives.Length <= LanLogger.MaxArchives,
            $"should not exceed MaxArchives, got {archives.Length}");

        if (File.Exists(LanLogger.LogPath))
        {
            var size = new FileInfo(LanLogger.LogPath).Length;
            Assert.IsTrue(size <= LanLogger.MaxBytes * 2,
                $"active log should be near the rotation cap, got {size} bytes");
        }
    }

    /// Logging to a non-writable path must never throw or crash.
    [TestMethod]
    public void LoggingToUnwritablePathNeverCrashes()
    {
        LanLogger.TestLogDirectoryOverride = @"Q:\nonexistent-drive\impossible\" + Guid.NewGuid();
        LanLogger._TestResetHeaderFlag();

        LanLogger.Info("Safe", "app");
        LanLogger.Warn("Safe", "warn");
        LanLogger.Error("Safe", "error");
        LanLogger.FileTransfer("start", transferId: "x", filename: "f.bin");
        LanLogger.Screenshot("captured", widthPx: 100, heightPx: 100);
        LanLogger.Peer("connect", peer: "127.0.0.1");
        LanLogger.Info("Discovery", "started interfaces=1");
        LanLogger.Info("Crypto", "key_generated algorithm=X25519");
        LanLogger.Info("UI", "window_shown");
        LanLogger.Info("Retry", "retry subsystem=Net attempt=1");
        // Pass = did not throw.
    }

    // ── Frame codec stress ────────────────────────────────────────────────────────

    /// Round-trips JSON frames at a range of content sizes through EncodeDict →
    /// MemoryStream → ReadFrame, including small and large (1 MiB) payloads.
    ///
    /// Note: FrameCodec encodes typed objects/dicts as JSON, not raw byte
    /// arrays. Size=0-byte JSON bodies are rejected by BuildFrame, so the
    /// smallest meaningful unit is a dict with a 1-char string value.
    [TestMethod]
    public void FrameCodecRoundTripAtBoundaries()
    {
        // String-value lengths to embed in the JSON frame body.
        int[] contentSizes = [1, 2, 127, 128, 255, 256, 1_024, 65_535, 65_536, 1_048_576];

        foreach (var size in contentSizes)
        {
            var content = new string('x', size);
            var dict    = new Dictionary<string, object?> { ["data"] = content };
            var encoded = FrameCodec.EncodeDict(dict);

            using var ms   = new MemoryStream(encoded);
            var       body = FrameCodec.ReadFrame(ms);

            Assert.IsNotNull(body,
                $"ReadFrame returned null for content size={size}");

            using var doc       = FrameCodec.ParseJson(body!);
            var       recovered = doc.RootElement.GetProperty("data").GetString();
            Assert.AreEqual(content, recovered,
                $"round-trip failed for content size={size}");
        }
    }

    /// Concatenates 5 000 back-to-back frames into one buffer and reads them
    /// all back via ReadFrame; verifies no frame is dropped or corrupted.
    [TestMethod]
    public void FrameCodecHighVolumeBackToBack()
    {
        const int frameCount = 5_000;
        var allBytes = new List<byte>();
        var expected = new List<string>();

        for (int i = 0; i < frameCount; i++)
        {
            var text = $"message {i} with some padding padding";
            allBytes.AddRange(FrameCodec.EncodeDict(
                new Dictionary<string, object?> { ["msg"] = text }));
            expected.Add(text);
        }

        var received = new List<string>();
        using var ms = new MemoryStream(allBytes.ToArray());
        byte[]? body;
        while ((body = FrameCodec.ReadFrame(ms)) is not null)
        {
            using var doc = FrameCodec.ParseJson(body);
            received.Add(doc.RootElement.GetProperty("msg").GetString()!);
        }

        Assert.AreEqual(frameCount, received.Count,
            "all frames must be received without loss");
        for (int i = 0; i < received.Count; i++)
            Assert.AreEqual(expected[i], received[i], $"frame {i} corrupted");
    }

    /// Verifies that frames whose declared length exceeds 50 MiB are rejected.
    [TestMethod]
    public void FrameCodecRejectsOversizeFrame()
    {
        uint tooBig = 50 * 1024 * 1024 + 1;
        var header = new byte[]
        {
            (byte)(tooBig >> 24),
            (byte)(tooBig >> 16),
            (byte)(tooBig >>  8),
            (byte)(tooBig      ),
        };

        using var ms = new MemoryStream(header);
        Assert.ThrowsException<InvalidDataException>(
            () => FrameCodec.ReadFrame(ms),
            "ReadFrame must reject frames whose declared size exceeds 50 MiB");
    }

    // ── Crypto stress ─────────────────────────────────────────────────────────────

    // Creates a fresh X25519 key with plaintext-export allowed (for test use only).
    private static Key MakeTempKey() => Key.Create(
        KeyAgreementAlgorithm.X25519,
        new KeyCreationParameters { ExportPolicy = KeyExportPolicies.AllowPlaintextExport });

    /// Encrypts and decrypts messages of varying sizes between a fixed key pair.
    [TestMethod]
    public void CryptoRoundTripHighVolume()
    {
        using var privA = MakeTempKey();
        using var privB = MakeTempKey();
        var pubAB64 = Convert.ToBase64String(privA.PublicKey.Export(KeyBlobFormat.RawPublicKey));
        var pubBB64 = Convert.ToBase64String(privB.PublicKey.Export(KeyBlobFormat.RawPublicKey));

        // SymmetricKey is the actual method name (not DeriveSymmetricKey).
        var keyA = SessionCrypto.SymmetricKey(privA, pubBB64);
        var keyB = SessionCrypto.SymmetricKey(privB, pubAB64);

        int[] sizes = [0, 1, 63, 64, 65, 1_024, 65_536, 500_000];

        foreach (var size in sizes)
        {
            var plaintext = new byte[size];
            for (int i = 0; i < size; i++) plaintext[i] = (byte)(i & 0xFF);

            var aad = Encoding.UTF8.GetBytes($"test-aad-{size}");
            var (nonceB64, ctB64) = SessionCrypto.Encrypt(keyA, plaintext, aad);
            var recovered = SessionCrypto.Decrypt(keyB, nonceB64, ctB64, aad);

            CollectionAssert.AreEqual(plaintext, recovered,
                $"round-trip failed for size={size}");
        }
    }

    /// Verifies that a tampered ciphertext is rejected.
    [TestMethod]
    public void CryptoRejectsTamperedCiphertext()
    {
        using var priv = MakeTempKey();
        var pubB64 = Convert.ToBase64String(priv.PublicKey.Export(KeyBlobFormat.RawPublicKey));
        var key    = SessionCrypto.SymmetricKey(priv, pubB64);
        var aad    = Encoding.UTF8.GetBytes("aad");

        var (nonceB64, ctB64) = SessionCrypto.Encrypt(key, Encoding.UTF8.GetBytes("hello"), aad);

        // Flip the last byte of the ciphertext (which is part of the GCM tag).
        var ctBytes = Convert.FromBase64String(ctB64);
        ctBytes[^1] ^= 0xFF;
        var tampered = Convert.ToBase64String(ctBytes);

        Assert.ThrowsException<System.Security.Cryptography.AuthenticationTagMismatchException>(
            () => SessionCrypto.Decrypt(key, nonceB64, tampered, aad),
            "tampered ciphertext must throw AuthenticationTagMismatchException");
    }

    /// Verifies that a wrong AAD is rejected.
    [TestMethod]
    public void CryptoRejectsWrongAAD()
    {
        using var priv = MakeTempKey();
        var pubB64 = Convert.ToBase64String(priv.PublicKey.Export(KeyBlobFormat.RawPublicKey));
        var key    = SessionCrypto.SymmetricKey(priv, pubB64);

        var (nonceB64, ctB64) = SessionCrypto.Encrypt(
            key, Encoding.UTF8.GetBytes("secret"), Encoding.UTF8.GetBytes("correct-aad"));

        Assert.ThrowsException<System.Security.Cryptography.AuthenticationTagMismatchException>(
            () => SessionCrypto.Decrypt(key, nonceB64, ctB64, Encoding.UTF8.GetBytes("wrong-aad")),
            "wrong AAD must be rejected by AES-GCM");
    }

    // ── Message-status regression ─────────────────────────────────────────────────

    /// Verifies that a late "Sent" notification cannot downgrade "Delivered" or "Read".
    ///
    /// MessageStatus constants are strings, not enum values. The comparison API is
    /// MessageStatus.Rank(s) and MessageStatus.ShouldApply(next, current) — there
    /// is no static Upgrade() helper.
    [TestMethod]
    public void MessageStatusNeverDowngrades()
    {
        var statuses = new[] { MessageStatus.Sent, MessageStatus.Delivered, MessageStatus.Read };

        foreach (var target in statuses)
        {
            foreach (var lower in statuses.Where(s => MessageStatus.Rank(s) < MessageStatus.Rank(target)))
            {
                // A lower-ranked status must not be allowed to overwrite a higher one.
                Assert.IsFalse(
                    MessageStatus.ShouldApply(lower, target),
                    $"status {target} must not downgrade to {lower}");
            }
        }
    }

    /// Verifies the upgrade path Sent → Delivered → Read is accepted,
    /// the reverse direction is rejected, and Read is idempotent.
    [TestMethod]
    public void MessageStatusUpgradeSequence()
    {
        // Forward (upgrade) transitions must be accepted.
        Assert.IsTrue(MessageStatus.ShouldApply(MessageStatus.Delivered, MessageStatus.Sent),
            "Delivered should apply over Sent");
        Assert.IsTrue(MessageStatus.ShouldApply(MessageStatus.Read, MessageStatus.Delivered),
            "Read should apply over Delivered");

        // Idempotent at the top.
        Assert.IsTrue(MessageStatus.ShouldApply(MessageStatus.Read, MessageStatus.Read),
            "Read → Read is idempotent and must be accepted");

        // Reverse (downgrade) transition must be rejected.
        Assert.IsFalse(MessageStatus.ShouldApply(MessageStatus.Sent, MessageStatus.Read),
            "Sent must not downgrade Read");
    }

    // ── Packet validator regression ───────────────────────────────────────────────

    /// Feeds malformed JSON blobs — validator must never crash.
    ///
    /// ValidateDiscovery and Validate (TCP) both take byte[], not string, and
    /// require senderIP / ownPublicKeyB64 / ownIPs parameters. There is no
    /// ValidateText method; the TCP validator is PacketValidator.Validate().
    [TestMethod]
    public void PacketValidatorHandlesMalformedInputWithoutCrash()
    {
        var inputs = new[]
        {
            "{}",
            "{\"type\":\"text\"}",
            "{\"type\":\"text\",\"message_id\":12345}",
            new string('a', 1_000),
            "{\"type\":\"discovery\",\"public_key\":\"\"}",
            "null",
            "[]",
            "",
            new string('{', 500),
        };

        // Dummy sender / own-key values that won't trigger self-suppression.
        const string ownKey = "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=";
        var          ownIPs = new HashSet<string>();

        for (int rep = 0; rep < 500; rep++)
        {
            foreach (var input in inputs)
            {
                var bytes = Encoding.UTF8.GetBytes(input);
                _ = PacketValidator.ValidateDiscovery(bytes, "1.2.3.4", ownKey, ownIPs);
                _ = PacketValidator.Validate(bytes, "1.2.3.4", ownKey);
            }
        }
        // Pass = no crash.
    }

    // ── History store regression ──────────────────────────────────────────────────

    /// Verifies that the history store caps at exactly 200 entries per peer.
    ///
    /// HistoryStore is a singleton (HistoryStore.Shared) with no public
    /// parameterised constructor and no per-call encryption key. The in-memory
    /// entry type is MessageEntry (not HistoryEntry), which uses Incoming:bool
    /// rather than a Direction enum and Timestamp:double (Unix seconds).
    /// The query method is Entries(peerIP), not Load(peer, key).
    [TestMethod]
    public void HistoryStoreCapsAt200Entries()
    {
        // Use a peer IP unlikely to collide with real or other-test data.
        var peer = $"10.99.{(Environment.TickCount & 0xFF)}.1";

        // Pre-clean any leftover state from a previous interrupted run.
        HistoryStore.Shared.Delete(peer);

        try
        {
            for (int i = 0; i < 250; i++)
            {
                var msg = new MessageEntry
                {
                    MessageId = i.ToString("x32"),
                    Sender    = "me",
                    Text      = $"message {i}",
                    Timestamp = DateTimeOffset.UtcNow.AddSeconds(i).ToUnixTimeSeconds(),
                    Status    = MessageStatus.Sent,
                    Incoming  = false,
                };
                HistoryStore.Shared.Append(msg, peer);
            }

            var entries = HistoryStore.Shared.Entries(peer);
            Assert.IsTrue(entries.Count <= 200,
                $"history must be capped at 200 entries, got {entries.Count}");
            Assert.IsTrue(entries.Count >= 190,
                $"history should keep at least 190 entries, got {entries.Count}");

            // Newest entry (249) must survive truncation.
            Assert.IsTrue(entries.Any(e => e.MessageId == (249).ToString("x32")),
                "newest entry must survive truncation");
        }
        finally
        {
            HistoryStore.Shared.Delete(peer);
        }
    }
}

File: src/macos/LanMessengerTests/StressTests.swift
import XCTest
import CryptoKit
@testable import LanMessenger

// Stress and regression tests for the LAN Messenger protocol stack.
//
// Goals
// -----
// • Verify that high-frequency concurrent operations (logging, crypto, frame
//   codec) produce correct results and never crash or dead-lock.
// • Verify that the rotation budget is maintained under continuous write load.
// • Catch common regressions: message-status downgrade, history truncation,
//   frame-codec round-trip correctness at boundary sizes.
//
// These tests run as part of `swift test` and in CI via pr-checks.yml and
// integration-test.yml.  They use no real sockets or disk state outside the
// temp-directory overrides.
final class StressTests: XCTestCase {

    // MARK: - Logger stress

    private var logTempDir: URL!
    private var savedMaxBytes: Int!
    private var savedMaxArchives: Int!

    override func setUp() {
        super.setUp()
        logTempDir = URL(fileURLWithPath: NSTemporaryDirectory())
            .appendingPathComponent("StressTests-\(UUID().uuidString)", isDirectory: true)
        try? FileManager.default.createDirectory(at: logTempDir, withIntermediateDirectories: true)
        NetLogger._testLogDirectoryOverride = logTempDir
        NetLogger._testResetHeaderFlag()
        savedMaxBytes    = NetLogger.maxBytes
        savedMaxArchives = NetLogger.maxArchives
    }

    override func tearDown() {
        NetLogger._testFlush()
        NetLogger.maxBytes    = savedMaxBytes
        NetLogger.maxArchives = savedMaxArchives
        NetLogger._testLogDirectoryOverride = nil
        NetLogger._testResetHeaderFlag()
        try? FileManager.default.removeItem(at: logTempDir)
        super.tearDown()
    }

    // 16 threads × 200 writes across all 8 channels = 3 200 total async writes.
    // Verifies thread-safety: no crashes, no interleaved partial lines.
    func testHighConcurrencyAcrossAllChannels() {
        NetLogger.maxBytes    = 512 * 1024   // 512 KiB — allow rotation mid-test
        NetLogger.maxArchives = 3

        let group   = DispatchGroup()
        let threads = 16
        let writes  = 200

        for t in 0..<threads {
            group.enter()
            DispatchQueue.global(qos: .userInitiated).async {
                for i in 0..<writes {
                    switch t % 8 {
                    case 0: NetLogger.info("Stress",     "t\(t) i\(i)")
                    case 1: NetLogger.fileTransfer(event: "chunk", transferId: "t\(t)",
                                                   bytesSent: Int64(i * 1024))
                    case 2: NetLogger.screenshot(event: "frame", widthPx: 1920, heightPx: 1080)
                    case 3: NetLogger.peer(event: "ping", peer: "10.0.\(t).1")
                    case 4: NetLogger.discovery(event: "beacon_sent", interfaces: t)
                    case 5: NetLogger.crypto(event: "derive", algorithm: "X25519")
                    case 6: NetLogger.ui(event: "frame_update", screen: "chat")
                    default: NetLogger.retry(event: "retry", subsystem: "Transfer", attempt: i)
                    }
                }
                group.leave()
            }
        }

        let timeout = DispatchTime.now() + .seconds(30)
        XCTAssertEqual(group.wait(timeout: timeout), .success,
                       "all writer threads should complete within 30 s")
        NetLogger._testFlush()

        // Every channel that received writes should have produced a non-empty file.
        for channel in NetLogger.LogChannel.allCases {
            let path = NetLogger._testLogURL(for: channel).path
            if FileManager.default.fileExists(atPath: path) {
                let size = (try? FileManager.default.attributesOfItem(atPath: path)[.size] as? Int) ?? 0
                XCTAssertGreaterThan(size, 0, "\(channel.logName) should not be empty")
            }
        }
    }

    // Continuously rotates a single channel; verifies archive count stays
    // within maxArchives and the active log never exceeds 2× the cap.
    func testRotationBudgetMaintainedUnderLoad() throws {
        NetLogger.maxBytes    = 256
        NetLogger.maxArchives = 2

        for i in 0..<500 {
            NetLogger.fileTransfer(event: "chunk", transferId: "rot",
                                   filename: "file_\(i).bin", size: Int64(i * 64))
        }
        NetLogger._testFlush()

        let dir      = logTempDir!
        let files    = (try? FileManager.default.contentsOfDirectory(atPath: dir.path)) ?? []
        let archives = files.filter { $0.hasPrefix("transfer.") && $0.hasSuffix(".log.gz") }

        XCTAssertLessThanOrEqual(archives.count, NetLogger.maxArchives,
            "should not exceed maxArchives; got \(archives.count)")

        let activeURL = NetLogger._testLogURL(for: .transfer)
        if FileManager.default.fileExists(atPath: activeURL.path) {
            let attrs = try FileManager.default.attributesOfItem(atPath: activeURL.path)
            let size  = attrs[.size] as? Int ?? 0
            XCTAssertLessThanOrEqual(size, NetLogger.maxBytes * 2,
                "active log should be near the rotation cap, got \(size) bytes")
        }
    }

    // Logging into a non-writable path must never throw or crash.
    func testLoggingToUnwritablePathNeverCrashes() {
        NetLogger._testLogDirectoryOverride = URL(fileURLWithPath: "/nonexistent/ro/\(UUID().uuidString)")
        NetLogger._testResetHeaderFlag()

        NetLogger.info("Safe", "app")
        NetLogger.warn("Safe", "warn")
        NetLogger.error("Safe", "error")
        NetLogger.fileTransfer(event: "start", transferId: "x", filename: "f.bin")
        NetLogger.screenshot(event: "captured", widthPx: 100, heightPx: 100)
        NetLogger.peer(event: "connect", peer: "127.0.0.1")
        NetLogger.discovery(event: "started", interfaces: 1)
        NetLogger.crypto(event: "key_generated", algorithm: "X25519")
        NetLogger.ui(event: "window_shown")
        NetLogger.retry(event: "retry", subsystem: "Net", attempt: 1)
        NetLogger._testFlush()
        // Pass = did not crash.
    }

    // MARK: - Frame codec stress

    // Helper Codable type for round-trip tests.
    private struct TestPacket: Codable, Equatable {
        let type: String
        let payload: String
        let seq: Int
    }

    // Round-trips many packets through the frame encoder + JSON parser.
    func testFrameCodecRoundTrip1000Packets() throws {
        for i in 0..<1_000 {
            let pkt = TestPacket(
                type: "stress",
                payload: String(repeating: "x", count: (i % 512)),
                seq: i
            )
            let frame  = try FrameCodec.encode(pkt)

            // Parse the frame back: strip the 4-byte length prefix.
            let bodyLength = Int(frame.prefix(4).withUnsafeBytes {
                $0.load(as: UInt32.self).byteSwapped
            })
            let body   = frame.dropFirst(4)
            XCTAssertEqual(bodyLength, body.count, "length prefix mismatch at seq=\(i)")

            let parsed = try FrameCodec.parseJSON(from: body)
            XCTAssertEqual(parsed["type"] as? String, "stress", "type mismatch at seq=\(i)")
            XCTAssertEqual(parsed["seq"] as? Int, i, "seq mismatch at i=\(i)")
        }
    }

    // Verifies that the encoder rejects a body that exceeds 50 MiB.
    // We test this by crafting a packet with a payload string large enough.
    func testFrameCodecRejectsOversizeBody() throws {
        // A 51 MiB string (each UTF-8 char is 1 byte for ASCII).
        let huge = TestPacket(type: "x", payload: String(repeating: "A", count: 51 * 1024 * 1024), seq: 0)
        XCTAssertThrowsError(try FrameCodec.encode(huge),
            "encoder must reject frames whose body exceeds 50 MiB")
    }

    // Verifies the length-prefix big-endian invariant across 500 packets.
    func testFrameCodecLengthPrefixIsBigEndianAlways() throws {
        for i in 0..<500 {
            let pkt   = TestPacket(type: "t", payload: "p\(i)", seq: i)
            let frame = try FrameCodec.encode(pkt)
            let hdr   = Array(frame.prefix(4))
            let declared = (UInt32(hdr[0]) << 24) | (UInt32(hdr[1]) << 16) |
                           (UInt32(hdr[2]) << 8)  |  UInt32(hdr[3])
            XCTAssertEqual(Int(declared), frame.count - 4,
                "big-endian length prefix wrong at seq=\(i)")
        }
    }

    // MARK: - Crypto stress

    // Encrypts and decrypts messages of varying sizes between two key pairs.
    func testCryptoRoundTripHighVolume() throws {
        let alice = Curve25519.KeyAgreement.PrivateKey()
        let bob   = Curve25519.KeyAgreement.PrivateKey()
        let bobPubB64   = bob.publicKey.rawRepresentation.base64EncodedString()
        let alicePubB64 = alice.publicKey.rawRepresentation.base64EncodedString()

        let messageSizes = [0, 1, 63, 64, 65, 512, 1_024, 10_000, 65_536]

        for size in messageSizes {
            var plaintext = Data(count: size)
            for i in 0..<size { plaintext[i] = UInt8(i & 0xFF) }

            let aad = Data("aad-\(size)".utf8)
            let (nonceB64, ctB64) = try SessionCrypto.encryptForPeer(
                myPrivate: alice, peerPublicKeyB64: bobPubB64, plaintext: plaintext, aad: aad)

            let recovered = try SessionCrypto.decryptFromPeer(
                myPrivate: bob, peerPublicKeyB64: alicePubB64,
                nonceB64: nonceB64, ciphertextB64: ctB64, aad: aad)

            XCTAssertEqual(recovered, plaintext, "round-trip failed for size=\(size)")
        }
    }

    // 200 concurrent encrypt/decrypt pairs must all succeed without races.
    func testCryptoConcurrentRoundTrips() throws {
        let alice = Curve25519.KeyAgreement.PrivateKey()
        let bob   = Curve25519.KeyAgreement.PrivateKey()
        let bobPubB64   = bob.publicKey.rawRepresentation.base64EncodedString()
        let alicePubB64 = alice.publicKey.rawRepresentation.base64EncodedString()

        let group     = DispatchGroup()
        let errLock   = NSLock()
        var cryptoErrors = [Error]()

        for i in 0..<200 {
            group.enter()
            DispatchQueue.global(qos: .userInitiated).async {
                defer { group.leave() }
                do {
                    let plaintext = Data("message \(i)".utf8)
                    let aad = Data("aad-\(i)".utf8)
                    let (nonceB64, ctB64) = try SessionCrypto.encryptForPeer(
                        myPrivate: alice, peerPublicKeyB64: bobPubB64, plaintext: plaintext, aad: aad)
                    let recovered = try SessionCrypto.decryptFromPeer(
                        myPrivate: bob, peerPublicKeyB64: alicePubB64,
                        nonceB64: nonceB64, ciphertextB64: ctB64, aad: aad)
                    if recovered != plaintext {
                        errLock.lock(); cryptoErrors.append(SessionCryptoError.decryptionFailed); errLock.unlock()
                    }
                } catch {
                    errLock.lock(); cryptoErrors.append(error); errLock.unlock()
                }
            }
        }

        let done = group.wait(timeout: .now() + .seconds(15))
        XCTAssertEqual(done, .success, "concurrent crypto round-trips timed out")
        XCTAssertTrue(cryptoErrors.isEmpty, "crypto errors in concurrent run: \(cryptoErrors)")
    }

    // Verifies that a tampered ciphertext is rejected (auth tag check).
    func testCryptoRejectsTamperedCiphertext() throws {
        let alice = Curve25519.KeyAgreement.PrivateKey()
        let bob   = Curve25519.KeyAgreement.PrivateKey()
        let bobPubB64   = bob.publicKey.rawRepresentation.base64EncodedString()
        let alicePubB64 = alice.publicKey.rawRepresentation.base64EncodedString()

        let (nonceB64, ctB64) = try SessionCrypto.encryptForPeer(
            myPrivate: alice, peerPublicKeyB64: bobPubB64,
            plaintext: Data("hello".utf8), aad: Data("aad".utf8))

        var ctData = Data(base64Encoded: ctB64)!
        ctData[ctData.count - 1] ^= 0xFF
        let tampered = ctData.base64EncodedString()

        XCTAssertThrowsError(
            try SessionCrypto.decryptFromPeer(
                myPrivate: bob, peerPublicKeyB64: alicePubB64,
                nonceB64: nonceB64, ciphertextB64: tampered, aad: Data("aad".utf8)),
            "tampered ciphertext must throw decryptionFailed")
    }

    // Verifies that a wrong AAD is rejected.
    func testCryptoRejectsWrongAAD() throws {
        let alice = Curve25519.KeyAgreement.PrivateKey()
        let bob   = Curve25519.KeyAgreement.PrivateKey()
        let bobPubB64   = bob.publicKey.rawRepresentation.base64EncodedString()
        let alicePubB64 = alice.publicKey.rawRepresentation.base64EncodedString()

        let (nonceB64, ctB64) = try SessionCrypto.encryptForPeer(
            myPrivate: alice, peerPublicKeyB64: bobPubB64,
            plaintext: Data("secret".utf8), aad: Data("correct-aad".utf8))

        XCTAssertThrowsError(
            try SessionCrypto.decryptFromPeer(
                myPrivate: bob, peerPublicKeyB64: alicePubB64,
                nonceB64: nonceB64, ciphertextB64: ctB64, aad: Data("wrong-aad".utf8)),
            "wrong AAD must be rejected by AES-GCM")
    }

    // MARK: - Message-status regression
    //
    // MessageStatus uses static string constants (not a Swift enum) with a
    // rank() function for monotonic-upgrade enforcement.

    // Verifies that a late "Sent" cannot downgrade "Delivered" or "Read".
    func testMessageStatusNeverDowngrades() {
        let statuses = [MessageStatus.sent, MessageStatus.delivered, MessageStatus.read]
        for target in statuses {
            for lower in statuses where MessageStatus.rank(lower) < MessageStatus.rank(target) {
                let should = MessageStatus.shouldApply(lower, over: target)
                XCTAssertFalse(should,
                    "shouldApply(\(lower as Any), over: \(target as Any)) must be false")
            }
        }
    }

    // Verifies the upgrade path: Sent → Delivered → Read.
    func testMessageStatusUpgradeSequence() {
        XCTAssertTrue(MessageStatus.shouldApply(MessageStatus.delivered, over: MessageStatus.sent),
            "Delivered must be accepted over Sent")
        XCTAssertTrue(MessageStatus.shouldApply(MessageStatus.read, over: MessageStatus.delivered),
            "Read must be accepted over Delivered")
        XCTAssertTrue(MessageStatus.shouldApply(MessageStatus.read, over: MessageStatus.read),
            "Same-rank transitions are allowed (idempotent)")
    }

    // Verifies the ranks are ordered as expected.
    func testMessageStatusRankOrder() {
        XCTAssertLessThan(MessageStatus.rank(MessageStatus.sent),
                          MessageStatus.rank(MessageStatus.delivered))
        XCTAssertLessThan(MessageStatus.rank(MessageStatus.delivered),
                          MessageStatus.rank(MessageStatus.read))
        XCTAssertLessThan(MessageStatus.rank(MessageStatus.failed),
                          MessageStatus.rank(MessageStatus.sent))
    }

    // MARK: - Packet validator regression

    // Feeds malformed JSON dicts — validator must never crash.
    func testPacketValidatorHandlesMalformedInputWithoutCrash() {
        let badDicts: [[String: Any]] = [
            [:],
            ["type": "text"],
            ["type": "text", "message_id": 12345],
            ["type": "discovery", "public_key": ""],
            ["type": "unknown_type_xyz"],
            ["type": "sent_receipt", "message_id": "abc"],
            ["type": NSNull()],
        ]

        let ownKey = "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA="

        for _ in 0..<300 {
            for dict in badDicts {
                _ = PacketValidator.validate(json: dict, senderIP: "10.0.0.1",
                                            ownPublicKeyB64: ownKey)
            }
        }
        // Pass = no crash.
    }

    // MARK: - History store — capacity regression

    // Verifies the maxEntriesPerPeer constant and the suffix logic that enforces it.
    func testHistoryCapLogicPreservesNewest() {
        let cap = HistoryStore.maxEntriesPerPeer
        XCTAssertEqual(cap, 200, "maxEntriesPerPeer should be 200 per PROTOCOL.md")

        // Simulate the .suffix(cap) truncation the store applies on save.
        let entries = (0..<250).map { i in "msg\(i)" }
        let kept    = Array(entries.suffix(cap))

        XCTAssertEqual(kept.count, cap,
            "suffix(\(cap)) should keep exactly \(cap) entries")
        XCTAssertEqual(kept.first, "msg50",
            "oldest surviving entry should be msg50")
        XCTAssertEqual(kept.last, "msg249",
            "newest entry must always be preserved")
    }
}
